src/consensus/versionbits.cpp

```diff
@@ -0,0 +1,179 @@
+
+#include "versionbits.h"
+
+using namespace Consensus::VersionBits;
+
+typedef std::multimap<int /*bit*/, int /*rule*/> RuleMap;
+typedef std::map<int /*rule*/, const SoftFork*> SoftForkMap;
+
+const char* Consensus::VersionBits::GetRuleStateText(int ruleState, bool bUseCaps)
+{
+    switch (ruleState)
+    {
+    case UNDEFINED:
+        return bUseCaps ? "UNDEFINED" : "undefined";
+
+    case DEFINED:
+        return bUseCaps ? "DEFINED" : "defined";
+
+    case LOCKED_IN:
+        return bUseCaps ? "LOCKED IN" : "locked in";
+
+    case ACTIVE:
+        return bUseCaps ? "ACTIVE" : "active";
+
+    case FAILED:
+        return bUseCaps ? "FAILED" : "failed";
+
+    default:
+        return bUseCaps ? "N/A" : "n/a";
+    }
+}
+
+bool Consensus::VersionBits::UsesVersionBits(int nVersion)
+{
+    return (nVersion & ~VERSION_BITS_MASK) == VERSION_HIGH_BITS;
+}
+
+
+SoftForkDeployments::~SoftForkDeployments()
+{
+    Clear();
+}
+
+void SoftForkDeployments::AddSoftFork(int bit, int rule, int threshold, uint32_t deployTime, uint32_t expireTime)
+{
+    if (bit < MIN_BIT || bit > MAX_BIT)
+        throw std::runtime_error("VersionBits::SoftForkDeployments::AddSoftFork() - invalid bit.");
+
+    if (threshold < 0)
+        throw std::runtime_error("VersionBits::SoftForkDeployments::AddSoftFork() - invalid threshold.");
+
+    if (deployTime >= expireTime)
+        throw std::runtime_error("VersionBits::SoftForkDeployments::AddSoftFork() - invalid time range.");
+
+    if (m_softForks.count(rule))
+        throw std::runtime_error("VersionBits::SoftForkDeployments::AddSoftFork() - rule already assigned.");
+
+    if (!IsBitAvailable(bit, deployTime, expireTime))
+        throw std::runtime_error("VersionBits::SoftForkDeployments::AddSoftFork() - bit conflicts with existing softFork.");
+
+    SoftFork* softFork = new SoftFork(bit, rule, threshold, deployTime, expireTime);
+    m_softForks.insert(std::pair<int, SoftFork*>(rule, softFork));
+    m_rules.insert(std::pair<int, int>(bit, rule));
+}
+
+bool SoftForkDeployments::IsBitAvailable(int bit, uint32_t deployTime, uint32_t expireTime) const
+{
+    std::pair<RuleMap::const_iterator, RuleMap::const_iterator> range;
+    range = m_rules.equal_range(bit);
+    for (RuleMap::const_iterator rit = range.first; rit != range.second; ++rit)
+    {
+        SoftForkMap::const_iterator it = m_softForks.find(rit->second);
+        if (it == m_softForks.end())
+            throw std::runtime_error("VersionBits::SoftForkDeployments::IsBitAvailable() - inconsistent internal state.");
+
+        // Do softFork times overlap?
+        const SoftFork* softFork = it->second;
+        if (((deployTime >= softFork->GetDeployTime()) && (deployTime <  softFork->GetExpireTime())) ||
+            ((expireTime >  softFork->GetDeployTime()) && (expireTime <= softFork->GetExpireTime())) ||
+            ((deployTime <= softFork->GetDeployTime()) && (expireTime >= softFork->GetExpireTime())))
+                return false;
```

Comment:
This seems to me that we can't reuse a bit for a new soft fork until the expire time, regardless of whether the previous soft fork at that bit has already activated.  I thought the intention was to allow reuse immediately after activation.